### PR TITLE
#2885 hasFractionalPart method added

### PR DESCRIPTION
**Pull request by siavashsoleymani**
This PR include the changes mentioned in the below issue
[#2885](https://github.com/FasterXML/jackson-databind/issues/2885)

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java [hunk 1]
```diff
@@ -57,6 +57,12 @@
     @Override
     public boolean isBigDecimal() { return true; }
 
+    @Override
+    public boolean hasFractionalPart() {
+        BigDecimal fractionalPart = _value.remainder( BigDecimal.ONE );
+        return !fractionalPart.equals(new BigDecimal("0").setScale(fractionalPart.scale()));
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java:63]: Use `BigDecimal.ZERO` to avoid constructing a new instance?

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java:63]: Actually, more specifically, create private `BigDecimal` instance to be used for comparison.

File: src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java [hunk 2]
```diff
@@ -52,6 +52,13 @@
     @Override
     public boolean isDouble() { return true; }
 
+    @Override
+    public boolean hasFractionalPart() {
+        long iPart = (long) _value;
+        double fPart = _value - iPart;
+        return fPart != 0.0;
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java:59]: Not sure if this works reliably (or needs Epsilon diff check) -- would probably good to check; if you have a link to somewhere that says this works reliably that'd be good. Not sure if coercion to `long` could lose also not work in some cases as `double` magnitude extends beyond `long` domain.

**Comment by cowtowncoder** [comment]
@siavashsoleymani Looks good overall; I have some questions on reliability of `double` / `float` part (and one suggestion for `BigDecimal` improvement), but will see if I can find answers to that.

One thing that would be nice would be bit more test coverage: there are couple of calls but I think it'd be good to try bit wider range of values to check edge cases for `float`/`double`/`BigDecimal`.

Other than this, the only thing I need is CLA from:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and the easiest way is usually to print it, fill & sign, scan/photo, email to `info` at fasterxml dot com.
This only needs to be done once (further contributions do not need another one); and once I receive CLA I can merge PR.

Looking forward to merging this!

File: src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java (hunk 1 above)
Comment (siavashsoleymani) [reply src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java:63]: Totally agree with you! as I know BigDecimal caches values from zero to ten. and I think the below code does the job and its not neccessary to create private BigDecimal instance:
`return !fractionalPart.equals(BigDecimal.ZERO.setScale(fractionalPart.scale()));`
am I correct?

**Comment by siavashsoleymani** [comment]
Thank you @cowtowncoder for your fast response and your great suggestions of course I will do and email it ASAP.

File: src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java (hunk 2 above)
Comment (siavashsoleymani) [reply src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java:59]: Good point thanks, let me investigate more on this.

File: src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java (hunk 1 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java:63]: Ah. I was hoping that `setScale()` could be avoided but looking at statement that is not possible.
But as per SO:

https://stackoverflow.com/questions/1078953/check-if-bigdecimal-is-integer-value

what can be done is something like this:

```java
private boolean isIntegerValue(BigDecimal bd) {
  return bd.signum() == 0 || bd.scale() <= 0 || bd.stripTrailingZeros().scale() <= 0;
}
```

which seems quite optimal. And in particular avoids `remainder()` which, being division, is probably quite expensive operation.

File: src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java (hunk 2 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java:59]: I think SO helps again: as per https://stackoverflow.com/questions/9898512/how-to-test-if-a-double-is-an-integer we could use `Math.rint()` like

```
if (x == Math.rint(x)) { // is integral
```

I am not 100% sure if NaN has to be checked separately; possibly not given that NaN typically is NOT the same is NaN (wrt Java `==`), but it might be fine to use full suggested comparison (from linked-to Guava impl https://guava.dev/releases/19.0/api/docs/com/google/common/math/DoubleMath.html#isMathematicalInteger(double))

```
!Double.isNaN(x) && !Double.isInfinite(x) && x == Math.rint(x)
```

... and since `DoubleNode` already defines `isNaN()` that does this, can just call that before or after using `Math.rint()`?

File: src/main/java/com/fasterxml/jackson/databind/node/FloatNode.java [hunk 3]
```diff
@@ -52,6 +52,13 @@
     @Override
     public boolean isFloat() { return true; }
 
+    @Override
+    public boolean hasFractionalPart() {
+        long iPart = (long) _value;
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/node/FloatNode.java:57]: Ok. Ideally we could do the same as with `double`, but surprisingly, `Math` has fewer method for `float` than `double`. So, two choices I think:

1. Cast to `double`, use `double` check.
2. Check for `isNaN()` of `FloatNode` first (NaNs need to considered to never be really integral, which probably means returning `true` for `hasFractionalPart()` (... but I now realize that name of method might not be optimal; I'll add separate note) and then use `if (_value == Math.round(_value)) { ... is integer, i.e. no fraction...)

**Comment by cowtowncoder** [comment]
Ok, cool, I think we can find relatively efficient methods to implement check for all relevant floating-point types. But now I wonder if naming makes sense -- I'll add my notes on issue itself, to be discussed separate from implementation details.

File: src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java (hunk 1 above)
Comment (siavashsoleymani) [reply src/main/java/com/fasterxml/jackson/databind/node/DecimalNode.java]: That's a great solution and more optimal! I will change the code to this one.

File: src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java (hunk 2 above)
Comment (siavashsoleymani) [reply src/main/java/com/fasterxml/jackson/databind/node/DoubleNode.java]: I change this part to something very simple and reliable, what do you think about the new one?

**Comment by siavashsoleymani** [comment]
Now I wonder about the right place for the hasFractionalPart method and I think based on the Interface segregation principle the right place for this method is in the NumericNode abstract class by that other types doesn't need anymore to implement or have this method by default.

**Comment by cowtowncoder** [comment]
@siavashsoleymani I can see why `NumericNode` might seem like the place, but I think `JsonNode` is better from actual usage perspective: less casting needed. The general division with `JsonNode` has been that read-only methods are added in `JsonNode` and only mutators (add, set, put) in subtypes. This allows all traversal without casts.

I think this is fine as-is, I can consider naming aspects after merging. So all I need now is CLA.

**Comment by siavashsoleymani** [comment]
@marcospereira Thank you for your kindness and explanations on this topic. The email has been sent.

**Comment by siavashsoleymani** [comment]
@cowtowncoder Thank you for your kindness and explanations on this topic. I learned a lot from you during this PR. Previous mention was wrong :)

**Comment by cowtowncoder** [comment]
@siavashsoleymani I guessed it might be. Thank you for working with me, it helped get this improvement out faster!
